### PR TITLE
Kp/regressor

### PR DESCRIPTION
**Pull request by kprokofi**
added async demo
bug fixes
metric per class added, eta added, other changes
new losses added
added optimization script

### CODE REVIEW

File: demo/demo.py [hunk 1]
```diff
@@ -145,13 +171,13 @@ def main():
...
-    parser.add_argument('--det_tresh', type=float, required=False, default=0.6)
+    parser.add_argument('--det_tresh', type=float, required=False, default=0.7)
     parser.add_argument('--device', type=str, default='CPU')
     parser.add_argument('-l', '--cpu_extension',
                         help='MKLDNN (CPU)-targeted custom layers.Absolute path to a shared library with the kernels '
                              'impl.', type=str, default=None)
-    parser.add_argument('--write_video', type=bool, default=False,
-                        help='if you set this arg to True, the video of the demo will be recoreded')
+    parser.add_argument('--write_video', action='store_true',
+                        help='wether or not to record demo video')
```
Comment (sovrasov) [inline demo/demo.py]: whether to save a demo video or not

File: torchdet3d/evaluation/metrics.py [hunk 2]
```diff
@@ -1,29 +1,43 @@
 import torch
 
-def compute_average_distance(pred_box, gt_box, num_keypoint=9):
+def compute_average_distance(pred_kp, gt_kp, num_keypoint=9, **kwargs):
     """Computes Average Distance (ADD) metric."""
-    detached_pred_box = pred_box.detach()
```
Comment (sovrasov) [inline torchdet3d/evaluation/metrics.py:13]: I propose keeping pytorch style: use torch.no_grad() when a gradient is not supposed to be calculated and don't call `detatch()` implicitly.

File: torchdet3d/losses/regression_losses.py [hunk 3]
```diff
@@ -58,6 +58,62 @@ def compute_diag(input_: torch.Tensor):
...
+                reg = reg_mean + reg_std
+            else:
+                cls = cls_mean
+                reg = reg_mean
+            if cls > reg:
+                self.lam_cls = (1 - (cls - reg)/cls).item()
+                print(f"classification coefficient changed : {self.lam_cls}")
+
+        return self.lam_reg * sum(regress_loss) + self.lam_cls * sum(class_loss)
+
```
Comment (sovrasov) [inline torchdet3d/losses/regression_losses.py]: I suggest `test()` moving this to the `tests` folder since it doesn't require test data.

File: torchdet3d/evaluation/metrics.py (hunk 2 above)
Comment (kprokofi) [reply torchdet3d/evaluation/metrics.py:13]: agree

File: torchdet3d/losses/regression_losses.py (hunk 3 above)
Comment (kprokofi) [reply torchdet3d/losses/regression_losses.py]: I can move all tests to tests folder and wrap it to unitests if needed. But how to be with dataloader tests.? It's required test data and path to root folder

Comment (sovrasov) [reply torchdet3d/losses/regression_losses.py]: Let's leave real data-related tests where they are for now.

**Comment by kprokofi** [comment]
@sovrasov , all done, what I wanted, please, review one more time

File: configs/default_config.py [hunk 4]
```diff
@@ -1,28 +1,29 @@
 data = dict(
     root="./data",
     resize=(224,224),
-    train_batch_size=64,
+    train_batch_size=128,
     val_batch_size=32,
-    max_epochs=200,
```
Comment (sovrasov) [inline configs/default_config.py:6]: Why here is only one epoch now?

File: configs/default_config.py [hunk 5]
```diff
@@ -1,28 +1,29 @@
...
-    num_workers=4,
+    max_epochs=1,
+    num_workers=6,
     category_list='all',
     normalization=dict(mean=[0.5931, 0.4690, 0.4229],
                        std=[0.2471, 0.2214, 0.2157])
 )
 
-model = dict(name='mobilenetv3_large', pretrained=True, num_classes=1, load_weights='')
+model = dict(name='mobilenetv3_large', pretrained=False, num_classes=9, load_weights='')
```
Comment (sovrasov) [inline configs/default_config.py]: Why `pretrained=False` by default?

File: configs/default_config.py [hunk 6]
```diff
@@ -1,28 +1,29 @@
...
 
-loss=dict(names=['mse', 'add_loss'], coeffs=([1., .1],[]), smoothl1_beta=0.2)
+loss=dict(names=['smoothl1',  'cross_entropy'], coeffs=([1.],[1.]), smoothl1_beta=0.2,
+                alwa=dict(use=False, lam_cls=1., lam_reg=1., C=100, compute_std=True))
 
-utils=dict(debug_mode = False, random_seeds=5, save_freq=10, print_freq=20, debug_steps=30)
+output_dir = './output/log'
 
-output_dir = './output/exp_0'
+utils=dict(debug_mode = False, random_seeds=5, save_freq=10, print_freq=20, debug_steps=100)
```
Comment (sovrasov) [inline configs/default_config.py]: Remove extra spaces: `debug_mode=False`

File: demo/demo.py [hunk 7]
```diff
@@ -12,12 +12,24 @@
 OBJECTRON_CLASSES = ('bike', 'book', 'bottle', 'cereal_box', 'camera', 'chair', 'cup', 'laptop', 'shoe')
 
 class Detector:
-    """Wrapper class for object detector"""
```
Comment (sovrasov) [inline demo/demo.py]: This was correct

File: demo/demo.py [hunk 8]
```diff
@@ -61,7 +73,7 @@ def __init__(self, ie,  model_path, device='CPU', ext_path=''):
     def get_detections(self, frame, detections):
         """Returns all detections on frame"""
         outputs = []
-        for rect in detections:
```
Comment (sovrasov) [inline demo/demo.py]: Let's leave it as is. It's hard to understand why do we need pick the first element here without any contest.

File: demo/demo.py [hunk 9]
```diff
@@ -12,12 +12,24 @@
...
         self.expand_ratio = (1., 1.)
 
+    def run_async(self, frame):
+        self.frame_shape = frame.shape
+        self.net.forward_async(frame)
+
+    def wait_and_grab(self):
+        outputs = self.net.grab_all_async()
+        detections = []
+        for out in outputs:
```
Comment (sovrasov) [inline demo/demo.py]: In this wrapper we suppose we have only one active infer request, se you can just add a check that length of `outputs` is one and don't run this loop at all.

File: torchdet3d/evaluation/evaluate.py [hunk 10]
```diff
@@ -27,6 +25,7 @@ class Evaluator:
     writer: object
     max_epoch: int
     device: str = 'cuda'
+    num_classes : int = 9
```
Comment (sovrasov) [inline torchdet3d/evaluation/evaluate.py]: ` = len(OBJECTRON_CLASSES)`
